rust/signed_doc/src/metadata/doc_type.rs

```diff
@@ -0,0 +1,225 @@
+                    )));
+                }
+
+                (0..len)
+                    .map(|_| parse_uuid(d))
+                    .collect::<Result<Vec<_>, _>>()
+                    .map(Self)
+                    .map_err(|e| {
+                        decode_context
+                            .report
+                            .other(&format!("Invalid UUIDv4 in array: {e}"), CONTEXT);
+                        minicbor::decode::Error::message(format!(
+                            "{CONTEXT}: Invalid UUIDv4 in array: {e}"
+                        ))
+                    })
+            },
+            minicbor::data::Type::Tag => {
+                // Handle single tagged UUID
+                match decode_context.conversion_policy {
+                    ConversionPolicy::Accept | ConversionPolicy::Warn => {
+                        if matches!(decode_context.conversion_policy, ConversionPolicy::Warn) {
+                            warn!("{CONTEXT}: Conversion of document type single UUID to type DocType");
+                        }
+
+                        parse_uuid(d).map(|uuid| Self(vec![uuid])).map_err(|e| {
+                            let msg = format!("Invalid single UUIDv4: {e}");
+                            decode_context.report.other(&msg, CONTEXT);
+                            minicbor::decode::Error::message(format!("{CONTEXT}: {msg}"))
+                        })
+                    },
+
+                    ConversionPolicy::Fail => {
+                        let msg = "Conversion of document type single UUID to type DocType is not allowed";
+                        decode_context.report.other(msg, CONTEXT);
+                        Err(minicbor::decode::Error::message(format!(
+                            "{CONTEXT}: {msg}"
+                        )))
+                    },
+                }
+            },
+            other => {
+                decode_context.report.invalid_value(
+                    "decoding type",
+                    &format!("{other:?}"),
+                    "array or tag cbor",
+                    CONTEXT,
+                );
+                Err(minicbor::decode::Error::message(format!(
+                    "{CONTEXT}: expected array of UUIDor tagged UUIDv4, got {other:?}",
+                )))
+            },
+        }
+    }
+}
+
+impl Encode<ProblemReport> for DocType {
+    fn encode<W: minicbor::encode::Write>(
+        &self, e: &mut minicbor::Encoder<W>, report: &mut ProblemReport,
+    ) -> Result<(), minicbor::encode::Error<W::Error>> {
+        const CONTEXT: &str = "DocType encoding";
+        e.array(self.0.len().try_into().map_err(|_| {
+            report.other("Unable to encode array length", CONTEXT);
+            minicbor::encode::Error::message(format!("{CONTEXT}, unable to encode array length"))
+        })?)?;
+        for id in &self.0 {
+            UuidV4::encode(id, e, &mut CborContext::Tagged).map_err(|_| {
```

Comment:
How can encoding a UUIDv4 actually fail?  Given its a type that is a uuidv4 already?